File: building-blocks/jpa/src/main/java/io/micro_spring_boot/building_blocks/jpa/domain/aggregate/AggregateRoot.java
package io.micro_spring_boot.building_blocks.jpa.domain.aggregate;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.domain.AbstractAggregateRoot;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.UUID;

@MappedSuperclass
@Getter(value = AccessLevel.PUBLIC)
@Setter(value = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public abstract class AggregateRoot<T extends AbstractAggregateRoot<T>> extends AbstractAggregateRoot<T> {
    public static final String SOFT_DELETED_CLAUSE = "deleted = false";

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(updatable = false)
    @CreatedBy
    private UUID createdBy;

    @Column(updatable = false)
    @CreatedDate
    private LocalDateTime createdAt;

    @Column
    @LastModifiedBy
    private UUID modifiedBy;

    @Column
    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Column(nullable = false, columnDefinition = "boolean default false")
    private boolean deleted;

    /**
     * Returns whether this aggregate root is new (meaning it has no ID) or persisted.
     */
    @Transient
    public boolean isNew() {
        return id == null;
    }
}

File: building-blocks/jpa/src/main/kotlin/com/microspringboot/building_blocks/jpa/domain/aggregate/AggregateRoot.kt
package com.microspringboot.building_blocks.jpa.domain.aggregate

import jakarta.persistence.*
import org.springframework.data.annotation.CreatedBy
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedBy
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.domain.AbstractAggregateRoot
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime
import java.util.*

@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
abstract class AggregateRoot<T : AbstractAggregateRoot<T>> : AbstractAggregateRoot<T>() {
    @Id
//    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(nullable = false, updatable = false)
    open var id: UUID = UUID.randomUUID()

    @CreatedBy
    open var createdBy: UUID? = null

    @LastModifiedBy
    var modifiedBy: UUID? = null

    @CreatedDate
    var createdAt: LocalDateTime? = null

    @LastModifiedDate
    var updatedAt: LocalDateTime? = null

    @Column(nullable = false, columnDefinition = "boolean default false")
    var isDeleted: Boolean = false
}